PsfLauncher/main.cpp

```diff
@@ -13,223 +13,227 @@
-        count = std::vsnprintf(str.data(), str.size() + 1, fmt, args2);
-        assert(count >= 0);
-        va_end(args2);
-    }
+    LPCWSTR ApplicationName;
+    std::wstring CommandLine;
+    LPCWSTR CurrentDirectory;
+};
+
+//Forward declarations
+ErrorInformation StartProcess(ExecutionInformation execInfo, int cmdShow, bool runInAppContainer);
+int launcher_main(PWSTR args, int cmdShow) noexcept;
+ErrorInformation RunScript(const psf::json_object * scriptInformation, std::filesystem::path packageRoot, LPCWSTR dirStr, int cmdShow);
+ErrorInformation GetAndLaunchMonitor(const psf::json_object * monitor, std::filesystem::path packageRoot, int cmdShow, LPCWSTR dirStr);
+ErrorInformation LaunchMonitorInBackground(std::filesystem::path packageRoot, const wchar_t * executable, const wchar_t * arguments, bool wait, bool asadmin, int cmdShow, LPCWSTR dirStr);
+static inline bool check_suffix_if(iwstring_view str, iwstring_view suffix);
+void LogStringW(const char* name, const wchar_t* value);
+void LogString(const char* name, const char* value);
+void Log(const char* fmt, ...);
+ErrorInformation StartWithShellExecute(std::filesystem::path packageRoot, std::filesystem::path exeName, std::wstring exeArgString, LPCWSTR dirStr, int cmdShow);
+bool CheckIfPowerShellIsInstalled();
 
-    str.resize(count);
-    ::OutputDebugStringA(str.c_str());
-}
-void LogString(const char* name, const char* value)
-{
-    Log("\t%s=%s\n", name, value);
-}
-void LogStringW(const char* name, const wchar_t* value)
+int __stdcall wWinMain(HINSTANCE, HINSTANCE, PWSTR args, int cmdShow)
 {
-    Log("\t%s=%ls\n", name, value);
+    return launcher_main(args, cmdShow);
 }
+
 int launcher_main(PWSTR args, int cmdShow) noexcept try
 {
+    //TEST
+    MessageBoxEx(NULL, L"In here", L"In here", 0, 0);
     Log("\tIn Launcher_main()");
     auto appConfig = PSFQueryCurrentAppLaunchConfig(true);
     if (!appConfig)
     {
         throw_win32(ERROR_NOT_FOUND, "Error: could not find matching appid in config.json and appx manifest");
     }
 
-    auto exeName = appConfig->get("executable").as_string().wide();
     auto dirPtr = appConfig->try_get("workingDirectory");
-    auto dirStr = dirPtr ? dirPtr->as_string().wide() : nullptr;
-
+    auto dirStr = dirPtr ? dirPtr->as_string().wide() : L"";
     auto exeArgs = appConfig->try_get("arguments");
-    auto exeArgString = exeArgs ? exeArgs->as_string().wide() : (wchar_t*)L"";
-    auto monitor = PSFQueryAppMonitorConfig();
-
 
     // At least for now, configured launch paths are relative to the package root
     std::filesystem::path packageRoot = PSFQueryPackageRootPath();
-    auto exePath = packageRoot / exeName;
 
-    // Allow arguments to be specified in config.json now also.
-    std::wstring cmdLine = L"\"" + exePath.filename().native() + L"\" " + exeArgString + L" " + args;
-
-    if (monitor != nullptr )
+    ErrorInformation error;
+    bool isPowershellInstalled = CheckIfPowerShellIsInstalled();

```

Comment:
the error reporting and returning seems kind of scattered.  consider making ErrorInformation and RAII that condidtionally calls PSFReportError on destruction.  Then your code either stashes error string/code or checks error code.

Reply:
What I'm trying to get here is to have one object that holds all the error information.  Thus the class ErrorInformation.  

I think the reason why it looks so scattered is because I promise that the ending script will run regardless if there is an exception, but only if the starting script ran.

Also, because I now have ErrorInformation, I needed a default ErrorInformation in case there was no error.

What is RAII?

Reply:
RAII: look it up  ;-)

I have no problem with the need for ErrorInfomation, just it's error-prone usage pattern, which RAII would help.

Reply:
https://osgwiki.com/wiki/WIL_Resource_Wrappers_(RAII)

(see reg handle comment below)